### PR TITLE
fix(space-tool): move and resize elements in correct order

### PR DESCRIPTION
**Pull request by philippfromme**
Related to bpmn-io/bpmn-js#1019

### CODE REVIEW

File: test/spec/features/space-tool/SpaceToolSpec.js [hunk 1]
```diff
@@ -637,6 +639,158 @@ describe('features/space-tool', function() {
   });
 
 
+  describe('steps', function() {
```
Comment (nikku) [inline test/spec/features/space-tool/SpaceToolSpec.js]: What is the actual scenarios that we fix with this new version of the tool? I presume there are some?

If that is the case, can we create full tests for these scenarios, too? Otherwise we'll not be able to verify what we fixed now is still working after the next rewrite of the space tool :tm: :rabbit: :copyright: .

File: lib/features/modeling/cmd/MoveShapeHandler.js [hunk 2]
```diff
@@ -60,26 +62,41 @@ MoveShapeHandler.prototype.postExecute = function(context) {
 
   var shape = context.shape,
       delta = context.delta,
-      hints = context.hints;
```
[the remark this answers is not in the document]
Comment (nikku) [reply lib/features/modeling/cmd/MoveShapeHandler.js]: Never mind, we have a test.

**Comment by philippfromme** [comment]
> we do not disable auto-resize during space tool

That shouldn't happen. I'll investigate.

File: lib/features/modeling/cmd/SpaceToolHandler.js [hunk 3]
```diff
@@ -81,7 +82,7 @@ SpaceToolHandler.prototype.updateConnectionWaypoints = function(
   forEach(connections, function(connection) {
     var source = connection.source,
         target = connection.target,
-        waypoints = connection.waypoints;
+        waypoints = copyWaypoints(connection);
```
Comment (nikku) [inline lib/features/modeling/cmd/SpaceToolHandler.js]: Why is this necessary?

Comment (philippfromme) [reply lib/features/modeling/cmd/SpaceToolHandler.js]: Mutation breaks undo/redo. By the way, this PR is work in progress.

File: lib/features/modeling/cmd/SpaceToolHandler.js [hunk 4]
```diff
@@ -1,49 +1,162 @@
...
+  var delta = context.delta,
+      direction = context.direction,
+      movingShapes = context.movingShapes,
       resizingShapes = context.resizingShapes,
-      delta = context.delta,
-      direction = context.direction;
+      start = context.start;
+
+  // (1) move shapes
+  this.moveShapes(movingShapes, delta);
```
Comment (nikku) [inline lib/features/modeling/cmd/SpaceToolHandler.js]: :heart:

**Comment by nikku** [comment]
Seems like there is a [significant amount of functionality](https://codecov.io/gh/bpmn-io/diagram-js/compare/553b617ce3d14c9818ef2eeae80329b8fad2f6fa...29e8f2e2eb886e90f94bd88b46ba3c3c0b2507ed/diff) not covered by tests.

**Comment by philippfromme** [comment]
> Seems like there is a [significant amount of functionality](https://codecov.io/gh/bpmn-io/diagram-js/compare/553b617ce3d14c9818ef2eeae80329b8fad2f6fa...29e8f2e2eb886e90f94bd88b46ba3c3c0b2507ed/diff) not covered by tests.

I know, I didn't manage to add all the necessary tests yesterday.

**Comment by nikku** [comment]
Looking into it right now.

**Comment by nikku** [comment]
:pray:

**Comment by nikku** [comment]
Considering the integration with bpmn-js there is a last remaining bug I see:

The label adjustment seems not to be disabled. This makes labels move twice still:

capture ViivTG_optimized

**Comment by nikku** [comment]
Issue: `SpaceTool#calculateAdjustment` has been removed.

It [is used in bpmn-js, tough](https://github.com/bpmn-io/bpmn-js/search?q=spaceTool.calculateAdjustments&unscoped_q=spaceTool.calculateAdjustments).

File: test/spec/features/space-tool/SpaceToolSpec.js (hunk 1 above)
Comment (nikku) [reply test/spec/features/space-tool/SpaceToolSpec.js]: I'll create integration tests in bpmn-js for some edge cases.

**Comment by nikku** [comment]
There is one last case that is not properly taken care of (yet). 

capture Eizn9J_optimized

Given there exist diagonal connections, the label does not move with the connection. This may break the label <> connection layout as shown in the screencapture above.

How big of a deal is it, though? Let's discuss :speech_balloon:.

**Comment by philippfromme** [comment]
> How big of a deal is it, though? Let's discuss 💬.

This is exactly the behavior I'd expect. Better no magic at all than no magic plus one exception where we don't even know if that's helping the user. This can happen with any connection.

no-magic

**Comment by nikku** [comment]
O.K. 

So let's not consider this a blocker then. Full :racehorse: ahead :postal_horn:.
